specification/inference/_types/Services.ts

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { TaskType } from '../_types/TaskType'
+import { UserDefinedValue } from '@spec_utils/UserDefinedValue'
+
+/**
+ * Configuration options when storing the model config
+ */
+export class ModelConfig {
+  /**
+   * The service type
+   */
+  service: string
+  /**
+   * Settings specific to the service
+   */
+  service_settings: ServiceSettings
+  /**
+   * Task settings specific to the service and model
+   */
+  task_settings: TaskSettings
+}
+
+/**
+ * Represents a model as returned by the GET API
+ */
+export class ModelConfigContainer {
```

Comment:
There's a lot of overlap between `ModelConfig` and `ModelConfigContainer`.

Do we need `task_type` and `model_id` here, considering that the user provided them in the request? If we remove them, then we can keep only `ModelConfig`.

If it's deemed necessary to repeat this information, then we should replace `service`, `service_settings` and `task_settings` with a single `config: ModelConfig` field.

Reply:
Thanks for the feedback @swallez, your comments have shown that the GET models API is not modelled correctly.

In future the GET inference model API will support wild card expansion. For example get all the text embedding  models:
```
GET _inference/text_embedding/*
```

To support this `GetModelResponse` should be an Array of `ModelConfigContainer` not a single value. Once we have an array of model configs it is clear why `model_id` and `task_type` are required in the response because `model_id` was not included in the request. The same applies to `task_type` if we were to ask for all models regardless of task the response should contain the task type.


> If it's deemed necessary to repeat this information, then we should replace service, service_settings and task_settings with a single config: ModelConfig field.

++ I like to reuse. How can I incorporate the fields from `ModelConfig` into `ModelConfigContainer` without having `ModelConfig` as a nested object? I want just the fields defined in `ModelConfig` to appear in `ModelConfigContainer`.

```
export class ModelConfigContainer {
  /**
   * The model Id
   */
  model_id: string
  /**
   * The model's task type
   */
  task_type: TaskType

  include all the fields defined in ModelConfig
}
```

Reply:
Thanks for the explanation regarding wildcards.

Regarding reuse, we can have `ModelConfigContainer extends ModelConfig` with additional fields to achieve what you want.

Now there's nothing wrong with nesting (composition). Some of the target languages for client libraries (e.g. Go) don't support inheritance and will "flatten" inheritance hierarchies and produce unrelated types. So if we want to allow roundtripping (e.g. scenarios like get a ModelConfig, tweak it and Post it back), composition makes it easier.

Reply:
👍 I pushed the `ModelConfigContainer extends ModelConfig` change.

The inability to round trip config is a common problem in Elasticsearch and it's quite wrong headed that you can't GET then PUT. I will ask for some more opinions as I'd prefer not to change the format and the nested config object makes it harder to read

Reply:
We looked at the round-tripping problem and realised there are other reason the prevent round-tripping anyway. The configuration contains secret settings such as an API key. The API key is included on PUT to create the model but because it is a secret it is never returned on GET. This means that you can't simply copy and paste the config as you need to insert the API key.

Our preference is to have the GET response contain flat fields as we think that is more readable than nesting the config part in another object.